# Find out profitable app profile for Google play and App store

In this project i tried to find out the popular apps on Google play and App store which will help the developers to understand the target market and build app accordingly.


## Open the data and explore
As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.

I have only used some sample data to minimize our time and resources. Below are the two data set i have woked with.

-  A data set containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018.
-  A data set containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017.

Let's open the data set and explore the field to understand the sample data

In [1]:
from csv import reader

## App store Data set
open_file = open ('AppleStore.csv')
read_file = reader (open_file)
ios = list (read_file)
ios_header = ios [0]
ios = ios [1:]

## Google play Data set
open_file = open ('googleplaystore.csv')
read_file = reader (open_file)
anroid = list (read_file)
anroid_header = anroid [0]
anroid = anroid [1:]    

To make it easier to explore the two data sets, i have first wrote a function named explore_data() that we can use repeatedly to explore rows in a more readable way. We'll also add an option for our function to show the number of rows and columns for any data set.

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))


print (ios_header)        
print ('\n')        
explore_data (ios,1,5,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16


We see that we have 7197 iOS app and 16 columns. Some of the useful columns for this analysis are 'track_name', 'currency', 'price', 'rating_count_tot', 'prime_genre'.  

In [3]:
print (anroid_header)        
print ('\n') 
explore_data (anroid,1,5,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 10841
Number of columns: 13


For Goole play we have 10841 google apps and 13 columns. Some of the columns that will be useful for our analysis are 'App', 'Category', 'Rating', 'Type', 'Price', 'Genres'.

## Deleting Wrong data

In [4]:
print (anroid_header)        # header
print ('\n') 
print (anroid [10472])       # incorrect row

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


The row 10472 is for the app 'Life Made WI-Fi Touchscreen Photo Frame' and we see that the rating is 19. But we know that the maximum rating can be 5. So we will delete this row.

In [7]:
print(len(anroid))
del anroid[10472] 
print(len(anroid))

10841
10840


## Deleting duplicate data

There are several duplicate data in Google play data. Like the application 'Box' has 3 duplicate entry.

In [8]:
for app in anroid:
    name = app [0]
    if name == 'Box':
        print (app)

['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']


In total, there are 1,181 cases where an app occurs more than once:

In [10]:
duplicate_apps = []
unique_apps = []

for app in anroid:
    name = app [0]
    if name in unique_apps:
        duplicate_apps.append (name)
    unique_apps.append (name)

print ('Number of duplicate apps: ', len (duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])

Number of duplicate apps:  1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


Removing duplicate entry
    - We will not remove the entry randomly
    - We will check for the reviews for each app and keep the highest one
    - First we create a directory with apps with max reviews.

In [11]:
reviews_max = {}

for app in anroid:
    name = app [0]
    n_reviews = float (app [3])
    
    if name in reviews_max and reviews_max [name] < n_reviews:
        reviews_max [name] = n_reviews
    
    if name not in reviews_max:
        reviews_max [name] = n_reviews
        
 
print ('Length of the dictionary is: ',  len (reviews_max))

Length of the dictionary is:  9659


We will remove duplicate rows from data set
    - Check if the apps reviws is same as reviews_max
    - If not delete the row else keep it.

In [12]:
android_clean = []
already_added = []

for app in anroid:
    name = app [0]
    n_reviews = float (app [3])
    
    if (reviews_max [name] == n_reviews) and (name not in already_added):
        android_clean.append (app)
        already_added.append (name)
        
print (len (android_clean))
    


9659


## Removing non English app

If we explore the data we will found that there are few apps which are not for English speaking audiance.



In [15]:
print(ios[813][1])
print(ios[6731][1])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


We are not much interested of this non English languages. So we will identify those data and remove it.

The below fuction will take string and will return True or False base on the    character available on that string

In [13]:
def is_English (string):
    
    for character in string:
        if ord (character) > 127:
            return False
    return True

print (is_English ('Instagram'))
print (is_English ('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print (is_English ('Docs To Go™ Free Office Suite'))

True
False
False


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.

To minimize the impact of data loss, we'll only remove an app if its name has more than three non-ASCII characters:

In [14]:
def is_English (string):
    
    non_ascii = 0
    
    for character in string:
        if ord (character) > 127:
            non_ascii += 1
        
    if non_ascii > 3:
        return False
    else:
        return True

print (is_English ('Docs To Go™ Free Office Suite'))
print (is_English ('中国語 AQリスニング'))

True
False


The function still may not 100% accurate and still can remove some English apps but still it is good enough to continue our analysis.

we will filter non English app using the above function

In [17]:
ios_english = []
android_english = []

for app in android_clean:
    name = app [0]
    if is_English (name):
        android_english.append(app)

for app in ios:
    name = app [1]
    if is_English (name):
        ios_english.append(app)
        
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)
print('\n')

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

So we left with 9614 android app and 6183 iOS app.


## Seperating Free apps

Our main objective of this project was working with apps that are free of cost. So we will seperate the apps accordingly.

In [18]:
free_ios = []
free_android = []

for app in android_english:
    price = app [7]
    if price == '0':
        free_android.append(app)

for app in ios_english:
    price = app [4]
    if price == '0.0':
        free_ios.append(app)
        
print (len(free_ios))
print (len(free_android))

3222
8864


Now we are left with 3222 iOS and 8864 anroid apps.

## Most common apps by Genre 

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

We will begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the  prime_genre column of the App Store data set, and the Genres and Category columns of the Google Play data set.

In [19]:
# Generate frequency tables that show percentages
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


# Display the percentages in a descending order
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Let's start by examining the frequency table for the prime_genre column of the App Store data set.

In [20]:
display_table(free_ios, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


More than 50% of the free apps are games and then entertainment and video. So in app store most popular apps are fun related.

Let's continue by examining the Genres and Category columns of the Google Play data set

In [21]:
display_table(free_android, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The output seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

This behavior is also confirmed by the frequency table we see for the Genres column:

In [22]:
display_table(free_android, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

## Finding most popular apps by Genre on App store

Below, we calculate the average number of user ratings per app genre on the App Store:

In [23]:
genres_ios = freq_table(free_ios, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    
    for app in free_ios:
        genre_app = app[-5]
        if genre_app == genre:
            n_rating = float (app[5])
            total += n_rating
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)
        

Utilities : 18684.456790123455
Navigation : 86090.33333333333
Travel : 28243.8
Music : 57326.530303030304
Sports : 23008.898550724636
Business : 7491.117647058823
Health & Fitness : 23298.015384615384
Book : 39758.5
Shopping : 26919.690476190477
Finance : 31467.944444444445
Food & Drink : 33333.92307692308
Photo & Video : 28441.54375
Social Networking : 71548.34905660378
Reference : 74942.11111111111
Productivity : 21028.410714285714
News : 21248.023255813954
Education : 7003.983050847458
Medical : 612.0
Catalogs : 4004.0
Lifestyle : 16485.764705882353
Games : 22788.6696905016
Weather : 52279.892857142855
Entertainment : 14029.830708661417


From the output we see that Navigation has the most avg user.

## Most Popular Apps by Genre on Google Play

Calculate the average number of installs per app genre for the Google Play data set.

In [24]:
categories_android = freq_table(free_android, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in free_android:
        category_app = app[1]
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

EVENTS : 253542.22222222222
FOOD_AND_DRINK : 1924897.7363636363
PERSONALIZATION : 5201482.6122448975
EDUCATION : 1833495.145631068
HEALTH_AND_FITNESS : 4188821.9853479853
BEAUTY : 513151.88679245283
GAME : 15588015.603248259
MAPS_AND_NAVIGATION : 4056941.7741935486
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
LIFESTYLE : 1437816.2687861272
FINANCE : 1387692.475609756
SHOPPING : 7036877.311557789
TRAVEL_AND_LOCAL : 13984077.710144928
AUTO_AND_VEHICLES : 647317.8170731707
DATING : 854028.8303030303
SOCIAL : 23253652.127118643
ENTERTAINMENT : 11640705.88235294
PHOTOGRAPHY : 17840110.40229885
LIBRARIES_AND_DEMO : 638503.734939759
NEWS_AND_MAGAZINES : 9549178.467741935
SPORTS : 3638640.1428571427
HOUSE_AND_HOME : 1331540.5616438356
PARENTING : 542603.6206896552
MEDICAL : 120550.61980830671
ART_AND_DESIGN : 1986335.0877192982
PRODUCTIVITY : 16787331.344927534
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
BUSINESS : 1712290.1474201474
FAMILY : 3695641.8198090

On average, communication apps have the most installs: 38,456,119. 